src/server/db_slice.cc

```diff
@@ -1338,4 +1361,57 @@ void DbSlice::ResetUpdateEvents() {
   events_.update = 0;
 }
 
+// todo: perhaps we need to limit the total number of entry in the tracking
+// table so that we have a way to limit the amount of memory used for
+// client tracking
+void DbSlice::TrackKeys(ConnectionContext* cntx, int32_t tid,
+                        const std::vector<std::string_view>& keys) {
+  DVLOG(2) << "Start tracking the following keys for thread ID: " << tid
+           << ", client ID: " << cntx->conn()->GetClientId();
+  for (auto key : keys) {
+    // std::string_view k = key;
+    std::string k{key.begin(), key.end()};
+    DVLOG(2) << "  " << k;
+    if (client_tracking_map_.find(k) == client_tracking_map_.end()) {
+      std::pair<ConnectionContext*, int32_t> p{cntx, tid};
+      absl::flat_hash_set<std::pair<ConnectionContext*, int32_t>, Hash> tracker_set{p};
+      client_tracking_map_.insert({k, tracker_set});
+    } else {
+      std::pair<ConnectionContext*, int32_t> p{cntx, tid};
+      client_tracking_map_[k].insert(p);
+    }
+  }
+}
+
+void DbSlice::SendInvalidationTrackingMessage(std::string_view key) {
+  std::string k{key.begin(), key.end()};
+  if (client_tracking_map_.find(k) != client_tracking_map_.end()) {
+    // notify all the clients.
+    auto& client_set = client_tracking_map_[k];
+    DVLOG(2) << "Garbage collect clients that are no longer tracking... ";
+    auto is_not_tracking = [](std::pair<ConnectionContext*, int32_t> p) {
+      return (!p.first->conn()->IsTrackingOn());
+    };
+    absl::erase_if(client_set, is_not_tracking);
+    DVLOG(2) << "Number of clients left: " << client_set.size();
+
+    if (!client_set.empty()) {
+      auto cb = [key, client_set](unsigned idx, util::ProactorBase*) {
+        for (auto it = client_set.begin(); it != client_set.end(); ++it) {
+          if ((unsigned int)it->second != idx)
+            continue;
+          facade::Connection* conn = it->first->conn();
+          DCHECK(conn);
+          facade::Connection::InvalidationMessage x{key};
+          conn->SendInvalidationMessageAsync(x);
+          return;
+        }
+      };
+      shard_set->pool()->DispatchBrief(std::move(cb));
```

Comment:
This is where you need to use the `Connection::WeakRef` because DispatchBrief just queues in a task to be dispatched, the pointer can be invalidated

Reply:
you meant iusing the weak reference nside the call back function `cb` right?